### PR TITLE
bump-allocator: document & poison

### PR DESCRIPTION
**Pull request by pepyakin**
In response to https://gitlab.parity.io/parity/vulnerability-response/-/issues/21

Please see each individual commit for details

### CODE REVIEW

File: primitives/allocator/src/freeing_bump.rs [hunk 1]
```diff
@@ -345,9 +368,12 @@ impl FreeingBumpHeapAllocator {
 
 				header_ptr
 			}
-			Link::Null => {
+			Link::Nil => {
 				// Corresponding free list is empty. Allocate a new item.
-				self.bump(order.size() + HEADER_SIZE, mem.size())?
+				Self::bump(
+					&mut self.bumper,
+					order.size() + HEADER_SIZE, mem.size()
```
Comment (bkchr) [inline primitives/allocator/src/freeing_bump.rs:375]:
```suggestion
					order.size() + HEADER_SIZE,
					mem.size(),
```

**Review by bkchr — APPROVED** [review]

File: primitives/allocator/src/freeing_bump.rs (hunk 1 above)
Comment (pepyakin) [reply primitives/allocator/src/freeing_bump.rs:375]: Oops 😅

**Comment by bkchr** [comment]
WAIT! Could we get a test? :D

**Comment by pepyakin** [comment]
yeah, sure, no rush. I am still doing my mini burnin meanwhile anyway

**Comment by bkchr** [comment]
But you actually did not change anything? Any error would have resulted in aborting the runtime anyway?

**Comment by pepyakin** [comment]
God saves man, who save himself

File: primitives/allocator/src/freeing_bump.rs [hunk 2]
```diff
@@ -851,4 +851,21 @@ mod tests {
...
+	#[test]
+	fn poison_oom() {
+		// given
+		// a heap of 32 bytes. Should be enough for two allocations.
+		let mut mem = [0u8; 32];
+		let mut heap = FreeingBumpHeapAllocator::new(0);
+
+		// when
+		assert!(heap.allocate(mem.as_mut(), 8).is_ok());
+		let alloc_ptr =heap.allocate(mem.as_mut(), 8).unwrap();
```
Comment (bkchr) [inline primitives/allocator/src/freeing_bump.rs:864]:
```suggestion
		let alloc_ptr = heap.allocate(mem.as_mut(), 8).unwrap();
```

**Review by bkchr — APPROVED** [review]

**Comment by pepyakin** [comment]
I've checked the build locally and it synced just fine, as expected. The test failure seems unrelated, looks like we need to bless the UI tests for the new rustc.

**Comment by pepyakin** [comment]
bot merge

**Pull request merged**